Fix ErrorDisplay so it shows error.message

LMS007 opened the pull request:
Fix ErrorDisplay to show the provided message in the rare cases where error.message is not empty AND the message prop is missing.

nb. In most cases error.message's text originated from the server.

--------

There are 4 possible states for the messages that ErrorDisplay renders

1. message && error.message = "message: _error.message_"
2. message && !error.message = "message"
3. !message && error.message = "error.message"
4. !message && !error.message = ""

Prior to this fix, case 3 was not working and so the error.message was not showing in the DOM.

run `make devdata`
Load up this assignment https://aunltd-test.blackboard.com/webapps/blackboard/execute/blti/launchPlacement?blti_placement_id=_20_1&content_id=_80_1&course_id=_19_1

...you can now see the error message in the dialog

_"A problem occurred while handling this request. Hypothesis has been notified."_

![Screen Shot 2021-06-23 at 3 06 33 PM](https://user-images.githubusercontent.com/3939074/123175640-f6de0500-d436-11eb-81bb-fec688dee243.png)

Fixes https://github.com/hypothesis/lms/issues/2859

LMS007 commented on lms/static/scripts/frontend_apps/components/test/LaunchErrorDialog-test.js:
```diff
@@ -83,7 +83,7 @@ describe('LaunchErrorDialog', () => {
       retryAction = 'Try again',
     }) => {
       it(`displays expected error for "${errorState}" error state`, () => {
-        const error = new Error('Detailed error info');
```
This now correctly gets printed to the DOM and affected this test. Simple fix is to just remove the error's message

robertknight commented on lms/static/scripts/frontend_apps/components/ErrorDisplay.js:
```diff
@@ -84,16 +86,20 @@ export default function ErrorDisplay({ message, error }) {
+      {!message && error.message && (
+        // Only the error's message property is provided
+        <p data-testid="message">{toSentence(error.message)}</p>
+      )}
```
I wonder if we ought to include a case for `!message && !error.message` here for the sake of completeness. This should almost never be triggered in practice because `Error`-like objects should almost always have a `message` property. This could perhaps use the string "An unknown error occurred."

LMS007 replied:
Sounds good.

robertknight reviewed (approved):
Thanks for the clear description + truth table of possibilities in the PR. I added some small suggestions but the main changes here look good.

LMS007 commented on lms/static/scripts/frontend_apps/components/test/ErrorDisplay-test.js:
```diff
@@ -151,6 +151,42 @@ describe('ErrorDisplay', () => {
+      );
+      if (output) {
+        assert.equal(
+          wrapper.find('p[data-testid="message"]').first().text(),
```
no, I think that was blind copy paste without thinking. Because I added data-testid, we don't the first() call.

LMS007 commented on lms/static/scripts/frontend_apps/components/ErrorDisplay.js:
```diff
@@ -84,16 +86,20 @@ export default function ErrorDisplay({ message, error }) {
-              {message}: <i>{toSentence(error.message)}</i>
-            </Fragment>
-          )}
+      {message && error.message && (
```
roger

LMS007 commented:
- Updated ErrorDisplayProps to explain in more detail how message and error.message work
- Added back exception string in LauchErrorDialog-test
- Added cases when error.message and message are both missing
- Addressed minor fixups in ErrorDisplay-test
